stdlib/camlinternalLazy.ml

```diff
@@ -46,15 +46,17 @@ let force_val_lazy_block (blk : 'arg lazy_t) =
 
 
 (* [force] is not used, since [Lazy.force] is declared as a primitive
-   whose code inlines the tag tests of its argument.  This function is
-   here for the sake of completeness, and for debugging purpose. *)
+   whose code inlines the tag tests of its argument, except when afl
+   instrumentation is turned on. *)
 
 let force (lzv : 'arg lazy_t) =
   let x = Obj.repr lzv in
   let t = Obj.tag x in
   if t = Obj.forward_tag then (Obj.obj (Obj.field x 0) : 'arg) else
   if t <> Obj.lazy_tag then (Obj.obj x : 'arg)
-  else force_lazy_block lzv
+  (* Use [Sys.opaque_identity] as [force_block_lazy] will modify its
+     argument, which could trigger warning 59 with flambda (see GPR#9998) *)
+  else force_lazy_block (Sys.opaque_identity lzv)
```

Comment:
I think we might prefer to do the `opaque_identity` earlier. Right now it is not a problem since `Obj.tag` is a C primitive, but we don't want this to break if this changes to be something that the compiler understand. If one of the previous if any one of those if is decided statically, it will break badly.
This probably hint that we also want to add `[@inline never]` as inlining this function might is never beneficial.

Reply:
That's what I did initially, but then I thought the other branches were pretty safe and we could benefit from the early cutting of branches if the tag is known. But I now remember that while it might be fine for the regular tags, it would be wrong for the forward tag. Would you accept a version where we first test for a regular tag, and in this case take the regular branch, otherwise wrap in `opaque_identity` then branch on either Lazy or Forward tag ?

Reply:
Never mind, I've tried it and it's too much complexity for very little gain, I'll push the initial version back again.

Reply:
The last version I pushed includes a comment with more details, does that help ?